### PR TITLE
[k8s public preview] Have EdgeDeploymentOperator report status to Agent

### PR DESCRIPTION
**Pull request by darobs**
Instead of allowing Agent to crash on Edge Deployment failure, the
failure that was reported to edge Deployment status will get reported
back in the EdgeAgent's twin. This more closely follows how single
device Edge Agent operates.

### CODE REVIEW

**Comment by darobs** [comment]
Reopening this to get it into beta11.  The EdgeDeploymentStatusCommandTest now confirms that the status message is *not* lost as @CindyXing 's request on #3315

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/edgedeployment/EdgeDeploymentOperatorTest.cs [hunk 1]
```diff
@@ -216,8 +215,7 @@ public async void StatusIsFailedWithSpecialMessageWhenHttpExceptionIsThrown()
                 client,
                 controller);
 
-            HttpOperationException ex = await Assert.ThrowsAsync<HttpOperationException>(() => edgeOperator.EdgeDeploymentOnEventHandlerAsync(WatchEventType.Added, edgeDefinition));
-            Assert.Equal(controllerException, ex);
+            await edgeOperator.EdgeDeploymentOnEventHandlerAsync(WatchEventType.Added, edgeDefinition);
             Assert.True(reportedStatus.HasValue);
             Assert.Equal(expectedStatus, reportedStatus.OrDefault());
```
Comment (CindyXing) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/edgedeployment/EdgeDeploymentOperatorTest.cs:220]: Given there is failure deploymentstatus, can we run the deploymentstatuscommand and verify? Same for the other one.

**Review by CindyXing — CHANGES_REQUESTED** [review]

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/edgedeployment/EdgeDeploymentOperatorTest.cs (hunk 1 above)
Comment (darobs) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/edgedeployment/EdgeDeploymentOperatorTest.cs:220]: That's more of an integration test, but I can add one.

Comment (CindyXing) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/edgedeployment/EdgeDeploymentOperatorTest.cs:220]: It's just we are eating exception. Or we can change the test case title.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentStatusCommand.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            // If the EdgeDeployment operator fails, report this to the plan executor
+            // as an exception.
+            this.activeStatus.ForEach(status =>
+                {
+                    if (status.State != EdgeDeploymentStatusType.Success)
+                    {
+                        throw new ConfigOperationFailureException(status.Message);
+                    }
+                });
```
Comment (zasherif) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentStatusCommand.cs:35]: This sounds like reducing all failures into one scenario and all what we expect the operator to do it to report the failure. I understand that the ReportDeploymentFailure method only reports the failure without an attempt to recover from it. Will it be helpful that the exception class characterize a failure type to propagate useful information about the failure? For example the Exception can have the following types:

1. TemporaryError; signals that the exception handler can retry the deployment
2. PermanentError; signals that the exception handler must permanently fail the deployment

That's also means that the plan executor should have a mechanism to retry (note: I did not read that part of the code yet).

**Review by zasherif — COMMENTED** [review]
lgtm, defer decisions to others.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentStatusCommand.cs (hunk 2 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentStatusCommand.cs:35]: > This sounds like reducing all failures into one scenario and all what we expect the operator to do it to report the failure

That is approximately how the single-device edge handles problems from command execution. Exceptions are caught and this is reduced to success or failure. However, C# has a notion of [fatal errors](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.devices.common.exceptionextensions.isfatal?view=azure-dotnet), and I should probably allow agent to crash if I get a fatal error. 

Overall, though, the idea of temporary errors vs. permanent may be useful - We have generally left it up to the plan runner to determine how often to retry commands, usually by progressively backing off, but usually the failures do land in one of these two categories.

**Review by zasherif — APPROVED** [review]
lgtm

**Pull request merged**